Python 3 support

pawelmhm opened the pull request:
~it's not ready for review but opening PR in case someone wants to contribute or help reviewing things.~ Edit, 10/02/2017 Tests should now pass in both python 3.5 an python 2.7,

chekunkov commented on scrapyrt/conf/__init__.py:
```diff
@@ -11,7 +11,7 @@ def __init__(self):
 
     def setmodule(self, module):
-        if isinstance(module, basestring):
+        if isinstance(module, str):
```
IMO it should be six.string_types

chekunkov commented on scrapyrt/core.py:
```diff
@@ -267,7 +267,8 @@ def create_spider_request(self, kwargs):
         except (TypeError, ValueError) as e:
-            message = "Error while creating Scrapy Request, {}".format(e.message)
+            msg = "Error while creating Scrapy Request, {}"
+            message = msg.format(str(e))
```
why not minimal change `e.message -> str(e)`?

pawelmhm replied:
this improves readability and doesn't do any harm

chekunkov replied:
ok, matter of taste

chekunkov commented on scrapyrt/log.py:
```diff
@@ -64,7 +64,7 @@ def _unicode_to_str(self, eventDict):
         if message:
             eventDict['message'] = tuple(
-                unicode_to_str(x, self.encoding) for x in message)
+                to_bytes(x, self.encoding) for x in message)
```
`to_bytes` is available only since Scrapy 1.1.0, ScrapyRT supports Scrapy>=1.0.0

pawelmhm replied:
this is good point, maybe we should add tests in tox for different scrapy versions what do you think?

chekunkov replied:
+1

pawelmhm replied:
added https://github.com/scrapinghub/scrapyrt/pull/49/commits/6dd934116e34a3dabcdd505f3cd1bd0d01d59f60

chekunkov commented on scrapyrt/resources.py:
```diff
@@ -74,9 +75,13 @@ def handle_error(self, exception_or_failure, request):
+        # Python exceptions don't have message attribute in Python 3+ anymore.
+        # Twisted HTTP Error objects still have 'message' attribute even in 3+
+        # and they fail on str(exception) call.
+        msg = exception.message if hasattr(exception, 'message') else str(exception)
```
exception in Python 2 and 3 have `.args` attr, you can use `exc.args[0]`. can you do that with `Twisted HTTP Error objects` in both py2 and py3?

pawelmhm replied:
Twisted HTTP Errors do have exc.args, but message is not always first index, for example sample 400 error can look like this:

```
In [2]: exception
Out[2]: twisted.web.error.Error('404', "'Spider not found: dmozx'", None)

In [3]: exception.args
Out[3]: ('404', "'Spider not found: dmozx'", None)
```
this means we can't rely on exc.args[0] refering to message.

pawelmhm replied:
and doing str(Twisted.web.error.Error) results in

```

TypeError: Can't convert 'bytes' object to str implicitly
```

chekunkov commented on tests/test_log_observer.py:
```diff
@@ -1,5 +1,5 @@
 # -*- coding: utf-8 -*-
-import StringIO
+from six import StringIO
```
in this case `six` is a direct dependency now

pawelmhm replied:
yes, good point

chekunkov commented on tests/test_resource_realtimeapi.py:
```diff
@@ -25,7 +25,7 @@ def _get_class_path(clsname):
 
     def test_realtimeapi_with_default_settings(self):
-        expected_entities = {'crawl.json': CrawlResource}
+        expected_entities = {b'crawl.json': CrawlResource}
```
are you sure it should be bytes all the way, not default str?

pawelmhm replied:
actually this is one thing where something may break for users and I'm little worried about this part, if route for resource is not bytes Twisted doesn't recognize resource path and returns 404 Not Found if you ask for crawl.json in Python 3. I made it bytes here, but actually maybe better approach would be to add encode() here https://github.com/scrapinghub/scrapyrt/blob/master/scrapyrt/resources.py#L100 this way if user has route that is not bytes but str it would still work

chekunkov replied:
+1 if you mean adding str support to `ServiceResource.putChild`

pawelmhm replied:
I'll just added to_bytes(route), so that before putChild route will be encoded to bytes.

chekunkov replied:
so these `b''` changes can be removed from tests now?

pawelmhm replied:
tests in `test_resource_realtime` are checking whether resource names are matched to resource object. Resource names must be bytes for Twisted, so we must keep b'' in tests. If resource name is string, not bytes it will return 404 in python 3. does it make sense?

pawelmhm replied:
aside fromt this thing ^ 3026199a9 responds to all other comments

pawelmhm commented:
I just added codecov to .travis.yml waiting for cow to post report here

pawelmhm commented:
codecov is silent, but looking into coverage report locally we have coverage on resources.py of around 65% so I'll add some more tests for lines that are not covered, to be 100% sure we dont introduce any regressins

pawelmhm commented:
@chekunkov porting goes smoothly, I added unit tests for parts of code that were not covered by unit tests (we had integration tests for resources but no unit tests). I also removed dead code here https://github.com/scrapinghub/scrapyrt/pull/49/commits/bf46bcfbebc42cb3a43b18b7bad89ed9454efabc it seems like this was not used anywhere and could be removed without problem. Now only part that misses tests is cmdline.py so I'll add tests for that. 
![coverage report](https://cloud.githubusercontent.com/assets/2700942/23313916/eb8aca58-fabf-11e6-852f-2d9f11d914d4.png)

pawelmhm commented on README.rst:
```diff
@@ -5,6 +5,15 @@ Scrapyrt (Scrapy realtime)
+
+.. image:: https://img.shields.io/pypi/l/scrapyrt.svg
+    :target: https://pypi.python.org/pypi/scrapyrt
+
```
this looks like this https://github.com/pawelmhm/scrapyrt/blob/6adc7ce91f70d65ecce9101bab12cfea1c4c5298/README.rst much better imo

chekunkov commented on tests/test_resource_serviceresource.py:
```diff
@@ -177,17 +178,17 @@ def test_render_object(self):
         self.assertEqual(
             self.resource.allowedMethods,
-            map(lambda s: s.strip(), access_control_allow_methods.split(','))
+            list(map(lambda s: s.strip(), access_control_allow_methods.split(',')))
```
maybe use list comprehension then?

huseyinyilmaz commented on scrapyrt/utils.py:
```diff
@@ -19,3 +23,23 @@ def extract_scrapy_request_args(dictionary, raise_error=False):
     return result
+
+
+def is_python2():
```
Since you have six as dependency, maybe you can use six.PY2 instead.
https://pythonhosted.org/six/#six.PY2

pawelmhm replied:
good point thanks, I fixed that
